File: app/src/main/java/com/openclassrooms/realestatemanager/injection/Injection.java
package com.openclassrooms.realestatemanager.injection;

import android.content.Context;

import com.openclassrooms.realestatemanager.data.local.database.RealEstateDatabase;
import com.openclassrooms.realestatemanager.data.local.reposiotries.ImageRepository;
import com.openclassrooms.realestatemanager.data.local.reposiotries.PropertiesRepository;
import com.openclassrooms.realestatemanager.data.viewModelFactory.ViewModelFactory;

import java.util.concurrent.Executor;
import java.util.concurrent.Executors;

public class Injection {

    public static PropertiesRepository sPropertiesRepository(Context context) {
        RealEstateDatabase db = RealEstateDatabase.getInstance(context);
        return new PropertiesRepository(db.singlePropertyDao());
    }

    public static ImageRepository sImageRepository(Context context) {
        RealEstateDatabase db = RealEstateDatabase.getInstance(context);
        return new ImageRepository(db.imageOfPropertyDao());
    }

    public static Executor provideExecutor(){ return Executors.newSingleThreadExecutor(); }

    public static ViewModelFactory sViewModelFactory(Context context) {
        PropertiesRepository propertiesRepository = sPropertiesRepository(context);
        ImageRepository imageRepository = sImageRepository(context);
        Executor executor = provideExecutor();
        return new ViewModelFactory(propertiesRepository, imageRepository, executor);
    }
}

File: app/src/main/java/com/openclassrooms/realestatemanager/injection/Injection.kt
package com.openclassrooms.realestatemanager.injection

import android.content.Context
import com.openclassrooms.realestatemanager.data.local.reposiotries.PropertiesRepository
import com.openclassrooms.realestatemanager.data.local.database.RealEstateDatabase
import com.openclassrooms.realestatemanager.data.local.reposiotries.ImageRepository
import com.openclassrooms.realestatemanager.data.remote.repository.GoogleMapsRepository
import com.openclassrooms.realestatemanager.data.viewModelFactory.ViewModelFactory
import java.util.concurrent.Executor
import java.util.concurrent.Executors

object Injection {

    private fun sPropertiesRepository(context: Context): PropertiesRepository {
        val db = context.let { RealEstateDatabase.getInstance(it) }
        return PropertiesRepository.getInstance(db?.singlePropertyDao())
    }

    private fun sImageRepository(context: Context): ImageRepository {
        val db = context.let { RealEstateDatabase.getInstance(it) }
        return ImageRepository.getInstance(db?.imageOfPropertyDao())
    }

    private fun sGoogleMapsRepository(): GoogleMapsRepository {
        return GoogleMapsRepository.getInstance()
    }

    private fun provideExecutor(): Executor {
        return Executors.newSingleThreadExecutor()
    }

    @JvmStatic
    fun sViewModelFactory(context: Context): ViewModelFactory {
        val propertiesRepository = sPropertiesRepository(context)
        val imageRepository = sImageRepository(context)
        val googleMapsRepository = sGoogleMapsRepository()
        val executor = provideExecutor()
        return ViewModelFactory(
            propertiesRepository,
            imageRepository,
            googleMapsRepository,
            executor
        )
    }
}